std/experimental/allocator/building_blocks/region.d

```diff
@@ -878,3 +939,388 @@ version(Posix) @system unittest
+    void[] alignedAllocate(size_t n, uint a)
+    {
+        import core.atomic : cas, atomicLoad;
+        import std.math : isPowerOf2;
+
+        assert(a.isPowerOf2);
+        if (n == 0) return null;
+
+        auto rounded = goodAllocSize(n);
+        shared void* localCurrent, localNewCurrent;
+
+        static if (growDownwards)
+        {
+            do
+            {
+                localCurrent = atomicLoad(_current);
+                auto alignedCurrent = cast(void*)(localCurrent - rounded);
+                localNewCurrent = cast(shared(void*)) alignedCurrent.alignDownTo(a);
+                if (alignedCurrent > localCurrent || localNewCurrent > alignedCurrent ||
+                    localNewCurrent < _begin)
+                    return null;
+            } while (!cas(&_current, localCurrent, localNewCurrent));
+
+            return cast(void[]) localNewCurrent[0 .. n];
+        }
+        else
+        {
+            do
+            {
+                localCurrent = atomicLoad(_current);
+                auto alignedCurrent = alignUpTo(cast(void*) localCurrent, a);
+                localNewCurrent = cast(shared(void*)) (alignedCurrent + rounded);
+                if (alignedCurrent < localCurrent || localNewCurrent < alignedCurrent ||
+                    localNewCurrent > _end)
+                    return null;
+            } while (!cas(&_current, localCurrent, localNewCurrent));
+
+            return cast(void[]) (localNewCurrent - rounded)[0 .. n];
+        }
+
+        assert(0, "Unexpected error in SharedRegion.alignedAllocate");
+    }
+
+    /**
+    Queries whether `b` has been allocated with this region.
+
+    Params:
+        b = Arbitrary block of memory (`null` is allowed; `owns(null)` returns
+        `false`).
+
+    Returns:
+        `true` if `b` has been allocated with this region, `false` otherwise.
+    */
+    pure nothrow @trusted @nogc
+    Ternary owns(const void[] b) const
+    {
+        return Ternary(b && (&b[0] >= _begin) && (&b[0] + b.length <= _end));
+    }
+
+    /**
+    Returns `Ternary.yes` if no memory has been allocated in this region,
+    `Ternary.no` otherwise. (Never returns `Ternary.unknown`.)
+    */
+    pure nothrow @safe @nogc
+    Ternary empty() const
+    {
+        import core.atomic : atomicLoad;
+
+        auto localCurrent = atomicLoad(_current);
+        static if (growDownwards)
+            return Ternary(localCurrent == _end);
+        else
+            return Ternary(localCurrent == _begin);
+    }
+
+    /**
+    If `ParentAllocator` is not $(REF_ALTTEXT `NullAllocator`, NullAllocator, std,experimental,allocator,building_blocks,null_allocator) and defines `deallocate`,
+    the region defines a destructor that uses `ParentAllocator.deallocate` to free the
+    memory chunk.
+    */
+    static if (!is(ParentAllocator == NullAllocator)
+        && hasMember!(ParentAllocator, "deallocate"))
+    ~this()
+    {
+        parent.deallocate(cast(void[]) _innerStorage);
```

Comment:
Looking at this, I don't think it is safe to assume that nobody else has a reference to `_innerStorage`, as we are passing a memory chunk allocated by the parent in `this(ubyte[])`.

Think of the following example:
```
auto buf = cast(ubyte[]) Mallocator.instance.allocate(42);
auto r1 = Region!(Mallocator)(buf);
auto r2 = Region!(Mallocator)(buf); // double free on ~this()
```

I believe we should take ownership of the memory chunk that gets passed into the region allocator, thus allowing us to safely deallocate the memory chunk.
In order to do this, we should define two `ctors` that take a memory chunk
* `this(ref ubyte[] store)` that takes an lvalue takes over the buffer
* `this(ubyte[] store)` that gets called by `this(size_t n)` which takes an rvalue

Another way would be to never `deallocate` a memory chunk that was passed to the region allocator, and only deallocate for regions that were created through `this(size_t n)`, thus leaving the resource management to the parent owning allocator.

What do you think?

Reply:
I'm not saying that the above sample of using buf is correct, but I don't think this should crash, though this is debatable.
A trickier use case, with an unexpected behavior would be
```
auto buf = cast(ubyte[]) Mallocator.instance.allocate(42);
{
  auto r = Region!(Mallocator)(buf);
  // do stuff with region
}
// attempt to reuse buf here and avoid allocating again
// and fail since Region silently deallocated buf
```